README.md
# teamr
You're having a cup for EA's popular football simulation, but no clue how to fairly assign teams to all players?
Sick of keeping notes up to date, to ensure a proper & fair team-roster?

Here comes **teamr**!

**teamr** takes the number of players, groups (e.g. when playing with many people) and the desired team-rating (absolute values and/or stars), giving you nice & fair groups for your tournament.

## Install

### Pre-compiled executables
Get them [here](http://github.com/hpbuniat/teamr/releases).

## Usage manual
```console
Usage: teamr [global options]

global options:
   --stars value, -s value       star selection mode (default: 0)
   --min value                   min. rating ovr, att, mid, def (default: 76)
   --max value                   max. rating ovr, att, mid, def (default: 80)
   --players value, -p value     number of players (default: 8)
   --groups value, -g value      number of groups (default: 2)
   --playerfile value, -f value  read players from file
   --nation, -n                  only national teams
   --women, -w                   only women national teams
   --help, -h                    show help
   --version, -v                 print the version
```
All selection-options are inclusive, so you will e.g. reset the ratings when using the star-selector only.

### Example
```console
teamr -p 8 -g 2 --stars 4 --min 0 --max 0
```
Will create 2 groups with 4 players each, using 4 star teams only
```console
====== Group 1
+-----------+----------------+-------------+-----+-----+-----+-----+-------+
|  PLAYER   |      TEAM      |   LEAGUE    | ATT | MID | DEF | OVR | STARS |
+-----------+----------------+-------------+-----+-----+-----+-----+-------+
| Player F  | Celta Vigo     | Liga BBVA   |  79 |  79 |  78 |  78 |   4.0 |
| Player A  | Olym. Lyonnais | Ligue 1     |  82 |  78 |  78 |  78 |   4.0 |
| Player D  | Chievo Verona  | Serie A TIM |  74 |  77 |  75 |  76 |   4.0 |
| Player H  | Milan          | Serie A TIM |  80 |  79 |  77 |  78 |   4.0 |
+-----------+----------------+-------------+-----+-----+-----+-----+-------+

====== Group 2
+-----------+-----------------+-------------+-----+-----+-----+-----+-------+
|  PLAYER   |      TEAM       |   LEAGUE    | ATT | MID | DEF | OVR | STARS |
+-----------+-----------------+-------------+-----+-----+-----+-----+-------+
| Player E  | FC Augsburg     | Bundesliga  |  75 |  76 |  74 |  76 |   4.0 |
| Player C  | Eint. Frankfurt | Bundesliga  |  77 |  73 |  75 |  75 |   4.0 |
| Player B  | West Brom       | Barclays PL |  78 |  76 |  75 |  76 |   4.0 |
| Player G  | 1. FC Köln      | Bundesliga  |  78 |  76 |  74 |  76 |   4.0 |
+-----------+-----------------+-------------+-----+-----+-----+-----+-------+
```

## Credits
- https://www.fifaindex.com/ for their incredible work of providing all the information teamr uses.


CHANGELOG.md
2016-12-20: v0.3.0
  - Add: read players from file
  - Add: re-assign a team for a single player

2016-12-20: v0.2.0
  - Fix group-assignment for unequal group-sizes
  - Improve parser to handle "None"

2016-12-13: v0.1.0
  - Initial release


scrapr_test.go
package main

import (
    "testing"
    "golang.org/x/net/html"
    "os"
    "reflect"
)

func prepare(file string) *html.Node {
    fixture, _ := os.Open("testdata/" + file + ".html")
    body, _ := html.Parse(fixture)
    return body
}

func TestParsr(t *testing.T) {
    fixture := prepare("p1")
    teams := []team {
        {"Olym. Lyonnais", "Ligue 1", 80, 78, 78, 78, 4.0},
        {"Milan", "Serie A TIM", 80, 78, 77, 78, 4.0},
        {"Zenit", "Russian League", 79, 80, 77, 78, 4.0},
        {"Lazio", "Serie A TIM", 79, 79, 78, 78, 4.0},
        {"Celta Vigo", "Liga BBVA", 79, 79, 78, 78, 4.0},
        {"West Ham", "Barclays PL", 79, 79, 76, 78, 4.0},
        {"Leicester City", "Barclays PL", 79, 77, 77, 78, 4.0},
        {"Fiorentina", "Serie A TIM", 79, 76, 79, 78, 4.0},
        {"Beşiktaş", "Süper Lig", 78, 78, 77, 78, 4.0},
        {"Real Sociedad", "Liga BBVA", 78, 78, 76, 78, 4.0},
        {"Southampton", "Barclays PL", 77, 77, 78, 78, 4.0},
        {"Shakhtar Donetsk", "Rest of World", 75, 79, 77, 78, 4.0},
        {"Crystal Palace", "Barclays PL", 81, 77, 75, 77, 4.0},
        {"Fenerbahçe", "Süper Lig", 79, 77, 78, 77, 4.0},
        {"OGC Nice", "Ligue 1", 79, 77, 78, 77, 4.0},
        {"Swansea City", "Barclays PL", 79, 77, 74, 77, 4.0},
        {"Real Betis", "Liga BBVA", 79, 76, 75, 77, 4.0},
        {"Torino", "Serie A TIM", 78, 78, 76, 77, 4.0},
        {"Watford", "Barclays PL", 77, 78, 77, 77, 4.0},
        {"Galatasaray", "Süper Lig", 77, 78, 74, 77, 4.0},
        {"Hertha BSC", "Bundesliga", 77, 77, 77, 77, 4.0},
        {"RCD Espanyol", "Liga BBVA", 77, 77, 77, 77, 4.0},
        {"1899 Hoffenheim", "Bundesliga", 77, 76, 76, 77, 4.0},
        {"Málaga CF", "Liga BBVA", 76, 77, 76, 77, 4.0},
        {"UD Las Palmas", "Liga BBVA", 75, 78, 76, 77, 4.0},
        {"Werder Bremen", "Bundesliga", 79, 75, 75, 76, 4.0},
        {"West Brom", "Barclays PL", 78, 76, 75, 76, 4.0},
        {"1. FC Köln", "Bundesliga", 78, 76, 74, 76, 4.0},
        {"Rubin Kazan", "Russian League", 78, 75, 73, 76, 4.0},
        {"Spartak Moscow", "Russian League", 77, 76, 76, 76, 4.0},
    }

    s := parsr(fixture)
    if (reflect.DeepEqual(s, teams) != true) {
        t.Errorf("parsr(%q) => %q, want %q", fixture, s, teams)
    }
}

func TestMore(t *testing.T) {
    var fixture = []struct {
        in  *html.Node
        out bool
    } {
        {
            prepare("p1"),
            true,
        }, {
            prepare("p2"),
            false,
        },
    }

    for _, sut := range fixture {
        s := more(sut.in)
        if s != sut.out {
            t.Errorf("more(%q) => %q, want %q", sut.in, s, sut.out)
        }
    }
}

func TestCreatr(t *testing.T) {
    var fixture = []struct {
        stars float64
        rateMin int
        rateMax int
        national bool
        women bool
        page int
        out string
    } {
        {
            0.0,
            76,
            80,
            false,
            false,
            1,
            "https://www.fifaindex.com/teams/1/?overallrating_0=76&overallrating_1=80&attackrating_0=76&attackrating_1=80&midfieldrating_0=76&midfieldrating_1=80&defenserating_0=76&defenserating_1=80&",
        }, {
            5.0,
            0,
            0,
            true,
            false,
            2,
            "https://www.fifaindex.com/teams/2/?type=1&stars=5",
        }, {
            4.5,
            0,
            0,
            false,
            true,
            2,
            "https://www.fifaindex.com/teams/2/?type=2&stars=4.5",
        },
    }

    for _, sut := range fixture {
        s := creatr(sut.stars, sut.rateMin, sut.rateMax, sut.national, sut.women, sut.page)
        if s.String() != sut.out {
            t.Errorf("creatr(%q, %q, %q, %q, %q, %q) => %q, want %q", sut.stars, sut.rateMin, sut.rateMax, sut.national, sut.women, sut.page, s.String(), sut.out)
        }
    }
}


drawr.go
package main

import (
    "fmt"
    "time"
    "math"
    "math/rand"
    "bufio"
    "os"
    "strings"
)

// playr collects the participating players, by reading either a text-file or querying them via stdin
func playr(number int, file string) []string {

    var players []string

    if (file != "") {
        file, err := os.Open(file)
        if err != nil {
            panic(err)
        }

        defer file.Close()

        scanner := bufio.NewScanner(file)
        for scanner.Scan() {
            t := strings.TrimSpace(scanner.Text())
            if (t != "") {
                players = append(players, t)
            }
        }
    } else {
        println("Enter player names (confirm with enter)")
        for i := int(0); i < number; i++ {
            reader := bufio.NewReader(os.Stdin)
            fmt.Printf("Player %d: ", i + 1)
            t, _ := reader.ReadString('\n')
            t = strings.TrimSpace(t)

            _ = "breakpoint"
            if (t != "") {
                players = append(players, t)
            }
        }
    }

    return players
}

// drawr does the actual drawing, by picking a random team for a random player.
// The result will be added to a group, starting with group 1, until the group is "filled".
func drawr(players []string, teams *[]team, groups int) map[int]draw {

    if (len(*teams) < len(players)) {
        fmt.Printf("Not enough teams found with given parameters (%d teams for %d players", len(*teams), len(players))
        os.Exit(1);
    }

    group := 1
    groupMax := int(math.Ceil(float64(len(players)) / float64(groups)))

    var drawn map[int]draw
    drawn = make(map[int]draw)
    drawn[group] = make(draw)

    for len(players) > 0 {
        rand.Seed(time.Now().UnixNano())
        if (len(drawn[group]) >= groupMax) {
            group++
            drawn[group] = make(draw)
        }

        player := rand.Int() % len(players)
        team := rand.Int() % len(*teams)

        drawn[group][players[player]] = (*teams)[team]

        // remove picked elements
        players[player] = players[len(players)-1]
        players = players[:len(players)-1]

        (*teams)[team] = (*teams)[len(*teams)-1]
        (*teams) = (*teams)[:len(*teams)-1]
    }

    return drawn
}


printr.go
package main

import (
    "os"
    "fmt"

    "github.com/olekukonko/tablewriter"
)

// printr prints the team assignment for all players per group
func printr(drawn map[int]draw) {
    for number, group := range drawn {
        table := tablewriter.NewWriter(os.Stdout)
        table.SetHeader([]string{"Player", "Team", "League", "Att", "Mid", "Def", "Ovr", "Stars"})

        for player, team := range group {
            row := []string{
                player,
                team.name,
                team.league,
                fmt.Sprintf("%d", team.att),
                fmt.Sprintf("%d", team.mid),
                fmt.Sprintf("%d", team.def),
                fmt.Sprintf("%d", team.ovr),
                fmt.Sprintf("%.1f", team.stars),
            }

            table.Append(row)
        }

        fmt.Printf("\n====== Group %d\n", number)
        table.Render()

        fmt.Print("\n====== Teams are ready!\n")
        fmt.Print("Confirm with enter or type a name to assign a new team (e.g. invalid teams according to more current stats\n\n")
    }
}


scrapr.go
package main

import (
    "net/http"

    "github.com/yhat/scrape"
    "golang.org/x/net/html"
    "golang.org/x/net/html/atom"
    "fmt"
    "net/url"
    "strings"
    "strconv"
)

// scrapr wraps all internal functions to collect all teams matching the search by querying fifaindex.com.
// The page is increased internally until there are no pages left.
func scrapr(stars float64, rateMin int, rateMax int, national bool, women bool) []team {

    var teamMap []team
    var body *html.Node
    page := 1

    for ok := true; ok; ok = more(body) {
        scrapeUrl := creatr(stars, rateMin, rateMax, national, women, page)
        body = readr(scrapeUrl)
        teamMap = append(teamMap, parsr(body)...)

        page++
    }

    return teamMap
}

// more is a internal helper to determine, if there are more pages to query
func more(body *html.Node) bool {
    matcher := func(n *html.Node) bool {
        if n != nil && n.DataAtom == atom.Li && n.Parent != nil && n.Parent.Parent != nil {
            return scrape.Attr(n, "class") == "next"
        }

        return false
    }

    return (body != nil && len(scrape.FindAll(body, matcher)) > 0)
}

// parsr does the actual parsing of teams & attributes.
// The result is a list of teams.
func parsr(body *html.Node) []team {
    teamRowMatcher := func(n *html.Node) bool {
        if n.DataAtom == atom.Tr && scrape.Attr(n, "class") == "" && n.Parent != nil && n.Parent.Parent != nil {
            return strings.Contains(scrape.Attr(n.Parent.Parent, "class"), "teams") && n.Parent.Parent.DataAtom == atom.Table
        }

        return false
    }

    teamAttrMatcher := func(n *html.Node) bool {
        if n.DataAtom == atom.Td && n.Parent != nil && n.Parent.Parent != nil {
            return n.Parent.DataAtom == atom.Tr && scrape.Attr(n, "data-title") != ""
        }

        return false
    }

    var teams []team
    teamRows := scrape.FindAll(body, teamRowMatcher)
    for _, teamRow := range teamRows {
        var teamContainer team
        teamAttrs := scrape.FindAll(teamRow, teamAttrMatcher)
        for _, teamAttr := range teamAttrs {

            title := scrape.Attr(teamAttr, "data-title")
            switch title {
                case "Name":
                    teamContainer.name = scrape.Text(teamAttr.FirstChild)

                case "League":
                    teamContainer.league = scrape.Text(teamAttr.FirstChild)

                case "ATT":
                    t, _ := strconv.ParseInt(scrape.Text(teamAttr.FirstChild), 10, 32)
                    teamContainer.att = int(t)

                case "MID":
                    t, _ := strconv.ParseInt(scrape.Text(teamAttr.FirstChild), 10, 32)
                    teamContainer.mid = int(t)

                case "DEF":
                    t, _ := strconv.ParseInt(scrape.Text(teamAttr.FirstChild), 10, 32)
                    teamContainer.def =  int(t)

                case "OVR":
                    t, _ := strconv.ParseInt(scrape.Text(teamAttr.FirstChild), 10, 32)
                    teamContainer.ovr = int(t)

                case "Team Rating":
                    teamContainer.stars = float32(len(scrape.FindAll(teamAttr, scrape.ByClass("fa-star")))) + (float32(len(scrape.FindAll(teamAttr, scrape.ByClass("fa-star-half-o")))) * float32(.5))
            }
        }

        // there is team with name "none" in the markup :)
        if (len(teamContainer.name) > 0 && teamContainer.name != "None") {

            teams = append(teams, teamContainer)
        }
    }

    return teams
}

// readr does the actual querying of fifaindex.com and parses the response-body into a html.Node
func readr(scrapeUrl *url.URL) *html.Node {
    println(fmt.Sprintf("Fetching: %s", scrapeUrl.String()))
    resp, err := http.Get(scrapeUrl.String())
    if err != nil {
        panic(err)
    }

    body, err := html.Parse(resp.Body)
    if err != nil {
        panic(err)
    }

    return body
}

// creatr builds the fully-qualified URL to query fifaindex.com
func creatr(stars float64, rateMin int, rateMax int, national bool, women bool, page int) *url.URL {

    baseUrl := fmt.Sprintf("https://www.fifaindex.com/teams/%d/?", page)
    if (rateMin > 0 && rateMax > 0) {
        baseUrl += fmt.Sprintf("overallrating_0=%d&overallrating_1=%d&", rateMin, rateMax)
        baseUrl += fmt.Sprintf("attackrating_0=%d&attackrating_1=%d&", rateMin, rateMax)
        baseUrl += fmt.Sprintf("midfieldrating_0=%d&midfieldrating_1=%d&", rateMin, rateMax)
        baseUrl += fmt.Sprintf("defenserating_0=%d&defenserating_1=%d&", rateMin, rateMax)
    }

    if national == true {
        baseUrl += "type=1&"
    } else if women == true {
        baseUrl += "type=2&"
    }

    if (stars > 0) {
        if (stars == float64(int(stars))) {
            baseUrl += fmt.Sprintf("stars=%.0f", stars)
        } else {
            baseUrl += fmt.Sprintf("stars=%.1f", stars)
        }
    }

    returnUrl, _ := url.Parse(baseUrl);
    return returnUrl
}


main.go
package main

import (
    "fmt"
    "os"

    "github.com/urfave/cli"
)

// team represents all necessary attributes of a football-team
type team struct {
    name string
    league string
    att int
    mid int
    def int
    ovr int
    stars float32
}

// draw allows the assignment of a team to each player
type draw map[string]team

// replace in build-process
var Version = "???"

// Enter the cli-program
func main() {
    var stars float64
    var rateMin int
    var rateMax int
    var national bool
    var women bool
    var players int
    var groups int
    var playerFile string

    app := cli.NewApp()
    app.Version = Version
    app.Name = "teamr"
    app.Usage = "football-cup team randomizer"
    app.Flags = []cli.Flag {
        cli.Float64Flag{
            Name:        "stars, s",
            Value:       0,
            Usage:       "star selection mode",
            Destination: &stars,
        },
        cli.IntFlag{
            Name:        "min",
            Value:       76,
            Usage:       "min. rating ovr, att, mid, def",
            Destination: &rateMin,
        },
        cli.IntFlag{
            Name:        "max",
            Value:       80,
            Usage:       "max. rating ovr, att, mid, def",
            Destination: &rateMax,
        },
        cli.IntFlag{
            Name:        "players, p",
            Value:       8,
            Usage:       "number of players",
            Destination: &players,
        },
        cli.IntFlag{
            Name:        "groups, g",
            Value:       2,
            Usage:       "number of groups",
            Destination: &groups,
        },
        cli.StringFlag{
            Name:        "playerfile, f",
            Value:       "",
            Usage:       "read players from file",
            Destination: &playerFile,
        },
        cli.BoolFlag{
            Name:        "nation, n",
            Usage:       "only national teams",
            Destination: &national,
        },
        cli.BoolFlag{
            Name:        "women, w",
            Usage:       "only women national teams",
            Destination: &women,
        },
    }

    app.Action = func(c *cli.Context) error {
        println(fmt.Sprintf("stars %.1f, min %d, max %d, nation %t, women %t", stars, rateMin, rateMax, national, women))
        teamMap := scrapr(stars, rateMin, rateMax, national, women)

        playerMap := playr(players, playerFile)

        // re-test amount of players, as one could enter an empty name
        players = len(playerMap)

        drawn := drawr(playerMap, &teamMap, groups)
        printr(drawn)
        for playerMap = playr(1, ""); len(playerMap) > 0; playerMap = playr(1, "") {
            drawn := drawr(playerMap, &teamMap, 1)
            printr(drawn)
        }

        return nil
    }

    app.Run(os.Args)
}
